### ISSUE TITLE
Some widgets are not being updated when switching enums back and forth

### ISSUE THREAD

**Issue by vladkorotnev** [Finnerale/druid-enums#8]
While the previous one was more of a nitpick, this one looks serious so we'd have to collaborate on that :-)

I'm currently using your solution in an embedded linux application for a device and it's been working great, however, I've noticed one thing that happens if you use it to switch views.

Basically, while updating from one enum element to the same one with different built-in data works, updating to a completely different one, and then the first one but with new data, shows the old data instead.

**Comment by vladkorotnev** [comment]
Nevermind the code I pasted, it was too hard to understand. Here's a sample you can `cargo run` and see what is going wrong.

[druid-enums-bug.zip](https://github.com/Finnerale/druid-enums/files/6154149/druid-enums-bug.zip)

p.s. this doesn’t include an example of using the Activation view directly without druid-enums, but I’ve tested it beforehand and the problem was not reproduced.

**Comment by luleyleo** [comment]
I'm a bit short on time right now, but after playing around a bit, I'm fairly sure it is a problem with `Label` and not this crate.
That would also explain why the QR widget works just fine. For example if you put an additional `bool` into `AppState::Boot` and display it in a checkbox on the waiting screen once with `("Test 1", false)` and once with `("Test 2", true)` the text will still be "Test1" the second time, but the checkbox will be checked. I hope that might help a bit.

**Comment by vladkorotnev** [comment]
Hm, that sounds interesting since it works fine when not using druid-enums library. I'm new to Rust, but so far it seems the whole library is mostly code-generation macros — if so, is it possible to somehow "expand" them into the source code? That would basically give a vanilla druid-only source, which could be used to investigate further.

I'll tinker around a bit more and report if I find anything!

**Comment by luleyleo** [comment]
You can use `cargo-expand` or its CLion integration to view the result of a proc-macro. But this is the output:

<details>
<summary>Click to expand</summary>

```rust
impl AppState {
    pub fn matcher() -> AppStateMatcher {
        AppStateMatcher::new()
    }
}

pub struct AppStateMatcher {
    boot: Option<
        ::druid::WidgetPod<((String, bool)), Box<dyn ::druid::Widget<((String, bool))>>>,
    >,
    wait_activation: Option<
        ::druid::WidgetPod<
            (screens::activation::InactiveTermInfo),
            Box<dyn ::druid::Widget<(screens::activation::InactiveTermInfo)>>,
        >,
    >,
    default_: Option<Box<dyn ::druid::Widget<AppState>>>,
    discriminant_: Option<::std::mem::Discriminant<AppState>>,
}

impl AppStateMatcher {
    pub fn new() -> Self {
        Self {
            boot: None,
            wait_activation: None,
            default_: None,
            discriminant_: None,
        }
    }
    pub fn default(mut self, widget: impl ::druid::Widget<AppState> + 'static) -> Self {
        self.default_ = Some(Box::new(widget));
        self
    }
    pub fn default_empty(mut self) -> Self {
        self.default_ = Some(Box::new(::druid::widget::SizedBox::empty()));
        self
    }
    pub fn boot(mut self, widget: impl ::druid::Widget<((String, bool))> + 'static) -> Self {
        self.boot = Some(::druid::WidgetPod::new(Box::new(widget)));
        self
    }
    pub fn wait_activation(
        mut self,
        widget: impl ::druid::Widget<(screens::activation::InactiveTermInfo)> + 'static,
    ) -> Self {
        self.wait_activation = Some(::druid::WidgetPod::new(Box::new(widget)));
        self
    }
}

impl ::druid::Widget<AppState> for AppStateMatcher {
    fn event(
        &mut self,
        ctx: &mut ::druid::EventCtx,
        event: &::druid::Event,
        data: &mut AppState,
        env: &::druid::Env,
    ) {
        if self.discriminant_ == Some(::std::mem::discriminant(data)) {
            match data {
                AppState::Boot(p0) => match &mut self.boot {
                    Some(widget) => widget.event(ctx, event, (p0), env),
                    None => (),
                },
                AppState::WaitActivation(p0) => match &mut self.wait_activation {
                    Some(widget) => widget.event(ctx, event, (p0), env),
                    None => (),
                },
            }
        }
    }
    fn lifecycle(
        &mut self,
        ctx: &mut ::druid::LifeCycleCtx,
        event: &::druid::LifeCycle,
        data: &AppState,
        env: &::druid::Env,
    ) {
        self.discriminant_ = Some(::std::mem::discriminant(data));
        if let ::druid::LifeCycle::WidgetAdded = event {
            if self.default_.is_none() && self.boot.is_none() {
                ::log::warn!("{}::{} variant of {:?} has not been set.", "AppStateMatcher", "boot", ctx.widget_id());
            }
            if self.default_.is_none() && self.wait_activation.is_none() {
                ::log::warn!("{}::{} variant of {:?} has not been set.", "AppStateMatcher", "wait_activation", ctx.widget_id());
            }
        }
        match data {
            AppState::Boot(p0) => match &mut self.boot {
                Some(widget) => widget.lifecycle(ctx, event, (p0), env),
                None => (),
            },
            AppState::WaitActivation(p0) => match &mut self.wait_activation {
                Some(widget) => widget.lifecycle(ctx, event, (p0), env),
                None => (),
            },
        }
    }
    fn update(
        &mut self,
        ctx: &mut ::druid::UpdateCtx,
        old_data: &AppState,
        data: &AppState,
        env: &::druid::Env,
    ) {
        match (old_data, data) {
            (AppState::Boot(old_p0), AppState::Boot(p0)) => match &mut self.boot {
                Some(widget) => widget.update(ctx, (p0), env),
                None => (),
            },
            (AppState::WaitActivation(old_p0), AppState::WaitActivation(p0)) => {
                match &mut self.wait_activation {
                    Some(widget) => widget.update(ctx, (p0), env),
                    None => (),
                }
            }
            _ => {
                ctx.children_changed();
            }
        }
    }
    fn layout(
        &mut self,
        ctx: &mut ::druid::LayoutCtx,
        bc: &::druid::BoxConstraints,
        data: &AppState,
        env: &::druid::Env,
    ) -> ::druid::Size {
        match data {
            AppState::Boot(p0) => match &mut self.boot {
                Some(widget) => {
                    let size = widget.layout(ctx, bc, (p0), env);
                    widget.set_layout_rect(ctx, (p0), env, size.to_rect());
                    size
                }
                None => bc.min(),
            },
            AppState::WaitActivation(p0) => match &mut self.wait_activation {
                Some(widget) => {
                    let size = widget.layout(ctx, bc, (p0), env);
                    widget.set_layout_rect(ctx, (p0), env, size.to_rect());
                    size
                }
                None => bc.min(),
            },
        }
    }
    fn paint(&mut self, ctx: &mut ::druid::PaintCtx, data: &AppState, env: &::druid::Env) {
        match data {
            AppState::Boot(p0) => match &mut self.boot {
                Some(widget) => widget.paint(ctx, (p0), env),
                None => (),
            },
            AppState::WaitActivation(p0) => match &mut self.wait_activation {
                Some(widget) => widget.paint(ctx, (p0), env),
                None => (),
            },
        }
    }
}
```

</details>

**Comment by luleyleo** [comment]
One more idea:
It might actually be `druid-enums` fault, because when the enum variant changes it only calls `ctx.children_changed` which as far as I know does not trigger an update for existing widgets. `CheckBox` and `QRView` might be fine with this because they only care about `paint` getting called which happens anyway. Fixing this requires tracking which widgets are already initialized and then just call `update` on those anyway. When I originally wrote this library, doing so was not possible, but since then I've added [`WidgetPod::is_initialized`](https://docs.rs/druid/0.7.0/druid/struct.WidgetPod.html#method.is_initialized) which would allow us to fix this issue (if that actually is the issue).

**Comment by vladkorotnev** [comment]
Yep, I've just tried calling `update` from there but it panics

```
thread 'main' panicked at 'assertion failed: self.old_data.is_none()', C:\Users\akasaka\.cargo\git\checkouts\druid-f71533f3d81c0bc8\17e03bd\druid\src\core.rs:952:17
```

**Comment by vladkorotnev** [comment]
Sorry, I was dumb. 

Adding the code as follows into the expanded form has fixed the issue:

```rust
_ => {
                ctx.children_changed();
                match data {
                    AppState::Boot(p0) => match &mut self.boot {
                        Some(widget) => if widget.is_initialized() { widget.update(ctx, (p0), env) },
                        None => (),
                    },
                    AppState::WaitActivation(p0) => match &mut self.wait_activation {
                        Some(widget) => if widget.is_initialized() { widget.update(ctx, (p0), env) },
                        None => (),
                    }
                }
            }
```

Now I just need to figure out how to add that into the macro and that should fix it :D

**Comment by luleyleo** [comment]
I've already done so with #9 :)

**Comment by vladkorotnev** [comment]
Thanks a lot for your help!

**Comment by vladkorotnev** [comment]
Sorry, that seems to have broken another thing, in particular — returning to a previous view mode.
If in the example above I replace the work thread segment with the following:

```rust
 std::thread::spawn(move || {
        c.wait();
        // -> Now displaying password 222
        std::thread::sleep_ms(3000);
        set_app_state(AppState::Boot("This worked, right? Next must be 333".to_string()));
        // -> Now displaying loading indicator
        std::thread::sleep_ms(3000);
        set_app_state(AppState::WaitActivation(screens::activation::InactiveTermInfo {
            nonce: String::from("PASSWORD 333")
        }));

        // -> Setting same data second time does not help too
        std::thread::sleep_ms(3000);
        set_app_state(AppState::Boot("The End".to_string()));
        // -> Now MUST display password 333, but displaying 222 on label and 333 on QR????
        std::thread::sleep_ms(3000);
        set_app_state(AppState::WaitActivation(screens::activation::InactiveTermInfo {
            nonce: String::from("PASSWORD 333")
        }));
    });
```

Then the screen just becomes stuck. Apparently it's due to the fact that the final state is the same as the one before the previous one (both have the string "PASSWORD 333" -- if I change it to something else then it works).

I wonder if it's somehow possible to force an update without rebuilding the whole hierarchy — I would assume the removed `children_changed` call is still needed, but either before or after the `update` call added in #9.

**Comment by vladkorotnev** [comment]
Sorry for the huge message amount, I fixed it in #10, now it works fine on my side of things.
